sw/device/tests/sim_dv/ast_clk_rst_inputs.c

```diff
@@ -0,0 +1,577 @@
+  CHECK_DIF_OK(dif_adc_ctrl_filter_set_enabled(
+      &adc_ctrl, kDifAdcCtrlChannel0, kDifAdcCtrlFilter0, kDifToggleEnabled));
+  CHECK_DIF_OK(dif_adc_ctrl_filter_set_enabled(
+      &adc_ctrl, kDifAdcCtrlChannel1, kDifAdcCtrlFilter0, kDifToggleEnabled));
+
+  if (first_adc_setup) {
+    CHECK_DIF_OK(dif_adc_ctrl_set_enabled(&adc_ctrl, kDifToggleEnabled));
+  }
+}
+
+void ast_enter_sleep_states_and_check_functionality(
+    dif_pwrmgr_domain_config_t pwrmgr_config,
+    dif_entropy_src_config_t entropy_src_config, uint32_t alert_idx) {
+  bool deepsleep;
+  int read_fifo_depth_val;
+  bool clock_check_status;
+  uint32_t unhealthy_fifos, errors, alerts;
+
+  const dif_edn_t edn0 = {
+      .base_addr = mmio_region_from_addr(TOP_EARLGREY_EDN0_BASE_ADDR)};
+  const dif_edn_t edn1 = {
+      .base_addr = mmio_region_from_addr(TOP_EARLGREY_EDN1_BASE_ADDR)};
+
+  if ((pwrmgr_config & (~kDifPwrmgrDomainOptionUsbClockInActivePower)) == 0) {
+    deepsleep = true;
+  } else {
+    deepsleep = false;
+  }
+
+  if (pwrmgr_testutils_is_wakeup_reason(&pwrmgr, 0)) {
+    CHECK_DIF_OK(dif_entropy_src_set_enabled(&entropy_src, kDifToggleDisabled));
+
+    const dif_entropy_src_fw_override_config_t fw_override_config = {
+        .entropy_insert_enable = true,
+        .buffer_threshold = kEntropyFifoBufferSize,
+    };
+
+    CHECK_DIF_OK(dif_entropy_src_fw_override_configure(
+        &entropy_src, fw_override_config, kDifToggleEnabled));
+
+    CHECK_DIF_OK(dif_entropy_src_configure(&entropy_src, entropy_src_config,
+                                           kDifToggleEnabled));
+
+    // Verify that the FIFO depth is non-zero via SW - indicating the reception
+    // of data over the AST RNG interface.
+    IBEX_SPIN_FOR(read_fifo_depth(&entropy_src) > 0, 1000);
+
+    // test recoverable event
+    test_event(alert_idx, kDifToggleDisabled, kAlertSet);
+
+    // Enable all the AON interrupts used in this test.
+    rv_plic_testutils_irq_range_enable(
+        &rv_plic, kTopEarlgreyPlicTargetIbex0,
+        kTopEarlgreyPlicIrqIdAdcCtrlAonMatchDone,
+        kTopEarlgreyPlicIrqIdAdcCtrlAonMatchDone);
+    CHECK_DIF_OK(dif_pwrmgr_irq_set_enabled(&pwrmgr, 0, kDifToggleEnabled));
+
+    // Enter low power mode.
+    LOG_INFO("Issued WFI to enter sleep.");
+    // Setup low power.
+    rstmgr_testutils_pre_reset(&rstmgr);
+
+    if (!deepsleep) {
+      // read fifo depth before enter sleep mode
+      read_fifo_depth_val = read_fifo_depth(&entropy_src);
+    }
+
+    // Configure ADC
+    adc_setup(first_adc_setup);
+
+    // set sleep mode
+    pwrmgr_testutils_enable_low_power(&pwrmgr, kDifPwrmgrWakeupRequestSourceTwo,
+                                      pwrmgr_config);
+
+    wait_for_interrupt();
+
+    // Interrupt should have been serviced.
+    CHECK(interrupt_serviced);
+
+    interrupt_serviced = false;
+
+  } else if (pwrmgr_testutils_is_wakeup_reason(
+                 &pwrmgr, kDifPwrmgrWakeupRequestSourceTwo)) {
+    if (deepsleep) {
```

Comment:
i'm really confused about this check. I thought our ROM enables entropy_src, shouldn't we see a non-zero after deep sleep?

Reply:
Should the entropy_src FIFO be filled after the ROM configurations? Is it true also for the test ROM?

According to https://github.com/lowRISC/opentitan/blob/master/sw/device/tests/entropy_src_ast_rng_req_test.c#L33-L56 after the boot the entropy is being configured, so we thought the entropy should be configured after the boot.

Reply:
so it's a bit of both.
The rom (both real and test) will enable the entropy complex in [bypass mode](https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/lib/testing/test_rom/test_rom_start.S;l=173-182?q=test_rom&ss=opentitan%2Fopentitan).  That should fill the entropy_src FIFO i think.  It might not look filled in test simply because it is taking some time.

Reply:
The FIFO is not filled even after waiting 5msec after the booter enables the entropy_src. Within how long the FIFO should be filled after the entropy_src is enabled?

Reply:
I double checked the entropy_src initialization in the ROM code, and it looks like the FIFO register for firmware is not enabled. So, in order for the FIFO to fill up, the entropy_src first needs to be configured accordingly.

It looks like this test configures `.entropy_insert_enable = true,` further below and then waits for the entropy FIFO to fill up with ` IBEX_SPIN_FOR(read_fifo_depth(&entropy_src) > 0, 1000);` - hence I think this should be ok.

Reply:
Thanks Michael. The question is if we are ok with the ROM code not enabling the FIFO register, but I guess this will be checked as part of the ROM checks.

Reply:
I believe this is ok as it has been reviewed as part of the ROM reviews - but we can double check with @alphan just to be sure.

Reply:
I think so too because we use the [`rnd` driver](https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/silicon_creator/lib/drivers/rnd.c;drc=94cdceaa94c19ce1876c03ef7d4c19f69c887c0a;l=20) during early boot (entropy_src -> ibex -> rnd_data register). @moidx do we need to change anything here?